File: LessonRunTime/Son.m
//
//  Son.m
//  LessonRunTime
//
//  Created by zhangdong on 16/5/20.
//  Copyright © 2016年 zhangdong. All rights reserved.
//

#import "Son.h"

@implementation Son
- (instancetype)init
{
    self = [super init];
    if (self) {
        
        NSLog(@"%@", NSStringFromClass([self class]));
        NSLog(@"%@", NSStringFromClass([super class]));
        
        // 都输出为Son
        // super 是告诉编译器，调用方法时，要去父类的方法中找，而不是本类中
    }
    return self;
}


@end

File: LessonRunTime/Son.cpp
/*
// @implementation Son

 // clang -rewrite-objc Son.m
 
static instancetype _I_Son_init(Son * self, SEL _cmd) {
    
    self = ((Son *(*)(__rw_objc_super *, SEL))(void *)objc_msgSendSuper)((__rw_objc_super){(id)self, (id)class_getSuperclass(objc_getClass("Son"))}, sel_registerName("init"));
    
    
    if (self) {
        NSLog( NSStringFromClass(((Class (*)(id, SEL))(void *)objc_msgSend)((id)self, sel_registerName("class"))));
        
        NSLog(NSStringFromClass(((Class (*)(__rw_objc_super *, SEL))(void *)objc_msgSendSuper)((__rw_objc_super){(id)self, (id)class_getSuperclass(objc_getClass("Son"))}, sel_registerName("class"))));

        // [self class]  id objc_msgSend(id self, SEL op, ...)
        // [super class] id objc_msgSendSuper(struct objc_super *super, SEL op, ...)
 
        
         struct objc_super {
         
            __unsafe_unretained id receiver;
            __unsafe_unretained Class super_class;
         
         };
 
 

    }
    return self;
}

*/
// @end
